### PR TITLE
Anurse/fix 2546 again

### PR DESCRIPTION
**Pull request by anurse**
Most of this is just unreverting #2718 (which broke the CI). I also had to fix up the Client Profile Framework Reference resolution logic.

/cc @davidfowl

### CODE REVIEW

**Comment by davidfowl** [comment]
spaces between your ifs and  ( :smile:

**Comment by anurse** [comment]
The travis PR build failure looks unrelated. I'll restart it.

File: src/Microsoft.Dnx.Runtime/FrameworkReferenceResolver.cs [hunk 1]
```diff
@@ -171,8 +171,11 @@ private static FrameworkInformation GetFrameworkInformation(FrameworkName target
 
             // Skip this on mono since it has a slightly different set of reference assemblies at a different
             // location
-            if (!RuntimeEnvironmentHelper.IsMono
-                && FrameworkDefinitions.TryPopulateFrameworkFastPath(targetFramework.Identifier, targetFramework.Version, referenceAssembliesPath, out frameworkInfo))
+            // Also, if the framework we're targetting has a profile (Client profile, basically), don't use the
+            // fast path because it will be wrong.
+            if (!RuntimeEnvironmentHelper.IsMono &&
+                string.IsNullOrEmpty(targetFramework.Profile) &&
+                FrameworkDefinitions.TryPopulateFrameworkFastPath(targetFramework.Identifier, targetFramework.Version, referenceAssembliesPath, out frameworkInfo))
```
Comment (davidfowl) [inline src/Microsoft.Dnx.Runtime/FrameworkReferenceResolver.cs]: We should pass this method the entire targetFramework and have it return false so we don't need to special case here.

File: test/Microsoft.Dnx.Runtime.Tests/DependencyManagement/FrameworkReferenceResolverFacts.cs [hunk 2]
```diff
@@ -55,17 +59,28 @@ public class FrameworkReferenceResolverFacts
...
-            Assert.Equal(Version.Parse(expectedVersion).ToString(), actualVersion.ToString());
+            if (string.IsNullOrEmpty(expectedPath))
+            {
+                Assert.False(result);
+            }
+            else
+            {
+                Assert.True(result);
+                Assert.Equal(Environment.ExpandEnvironmentVariables(expectedPath).Replace("REFASMSROOT", FrameworkReferenceResolver.GetReferenceAssembliesPath()), actualPath);
+                // Having this be Version->Version equality caused some problems...
```
Comment (davidfowl) [inline test/Microsoft.Dnx.Runtime.Tests/DependencyManagement/FrameworkReferenceResolverFacts.cs]: What problems?

File: test/Microsoft.Dnx.Tooling.FunctionalTests/DnuPackTests.cs [hunk 3]
```diff
@@ -1,580 +1,43 @@
...
-                Assert.NotEqual(0, exitCode);
-                Assert.NotEmpty(stdErr);
-                var unresolvedDependencyErrorCount = stdErr
-                    .Split(new[] { Environment.NewLine }, StringSplitOptions.RemoveEmptyEntries)
-                    .Where(line => line.Contains("The dependency NonexistentPackage >= 1.0.0 could not be resolved"))
-                    .Count();
-                Assert.Equal(1, unresolvedDependencyErrorCount);
-            }
+            sdk.Dnu.Restore(projectDir).EnsureSuccess();
+            sdk.Dnu.Pack(projectDir, outputDir).EnsureSuccess();
```
Comment (davidfowl) [inline test/Microsoft.Dnx.Tooling.FunctionalTests/DnuPackTests.cs]: This should have an assert.

**Comment by davidfowl** [comment]
Minor changes then :shipit:

File: test/Microsoft.Dnx.Runtime.Tests/DependencyManagement/FrameworkReferenceResolverFacts.cs (hunk 2 above)
Comment (anurse) [inline test/Microsoft.Dnx.Runtime.Tests/DependencyManagement/FrameworkReferenceResolverFacts.cs]: Implicit operators caused weird behavior. This is old stuff though, not new in this PR.

File: test/Microsoft.Dnx.Tooling.FunctionalTests/DnuPackTests.cs (hunk 3 above)
Comment (anurse) [inline test/Microsoft.Dnx.Tooling.FunctionalTests/DnuPackTests.cs]: `EnsureSuccess` isn't an assert?

Comment (davidfowl) [inline test/Microsoft.Dnx.Tooling.FunctionalTests/DnuPackTests.cs]: Nope, it just throws.

Comment (anurse) [inline test/Microsoft.Dnx.Tooling.FunctionalTests/DnuPackTests.cs]: Aka, an assert. Assert just throws too.

File: src/Microsoft.Dnx.Runtime/FrameworkReferenceResolver.cs (hunk 1 above)
Comment (anurse) [inline src/Microsoft.Dnx.Runtime/FrameworkReferenceResolver.cs]: :+1:

**Pull request closed without merging**